File: results/2026-05-06_173435/tasks/11-semantic-version-bumper/typescript-bun-sonnet/generated-code/bumper.ts
// bumper.ts
// Semantic version bumper: reads commits (conventional format), determines
// bump type (feat->minor, fix->patch, breaking->major), updates version file,
// and generates a CHANGELOG entry.

import { readFileSync, writeFileSync, existsSync } from "fs";
import { join } from "path";

// --- Types ---

export interface Commit {
  hash: string;
  type: string;
  scope?: string;
  message: string;
  breaking: boolean;
}

export type BumpType = "major" | "minor" | "patch" | "none";

export interface BumpResult {
  oldVersion: string;
  newVersion: string;
  bumpType: BumpType;
  commits: Commit[];
  changelog: string;
}

// --- Core parsing ---

// Parses a single conventional commit line:
// Format: <hash> <type>[(<scope>)][!]: <message>
export function parseCommit(line: string): Commit {
  const parts = line.match(/^([^\s]+)\s+(.+)$/);
  if (!parts) {
    return { hash: "", type: "unknown", message: line, breaking: false };
  }

  const hash = parts[1];
  const rest = parts[2];

  // BREAKING CHANGE in message body or ! after type/scope signals a major bump
  const hasBreakingMarker = /^[^:]+!:/.test(rest);
  const hasBreakingMessage = /breaking change/i.test(rest);
  const breaking = hasBreakingMarker || hasBreakingMessage;

  // type(scope)!: message  OR  type!: message  OR  type: message
  const typeMatch = rest.match(/^(\w+)(?:\(([^)]+)\))?!?:\s*(.+)$/);
  if (!typeMatch) {
    return { hash, type: "unknown", message: rest, breaking };
  }

  return {
    hash,
    type: typeMatch[1],
    scope: typeMatch[2] ?? undefined,
    message: typeMatch[3],
    breaking,
  };
}

// --- Bump type determination ---

// Precedence: major > minor > patch > none
export function determineBumpType(commits: Commit[]): BumpType {
  if (commits.length === 0) return "none";
  if (commits.some((c) => c.breaking)) return "major";
  if (commits.some((c) => c.type === "feat")) return "minor";
  if (commits.some((c) => c.type === "fix")) return "patch";
  return "none";
}

// --- Version manipulation ---

function parseVersion(version: string): [number, number, number] {
  const clean = version.replace(/^v/, "");
  const parts = clean.split(".").map(Number);
  if (parts.length !== 3 || parts.some((n) => isNaN(n))) {
    throw new Error(`Invalid semver string: "${version}"`);
  }
  return [parts[0], parts[1], parts[2]];
}

export function bumpVersion(version: string, bumpType: BumpType): string {
  if (bumpType === "none") return version;
  const [major, minor, patch] = parseVersion(version);
  switch (bumpType) {
    case "major": return `${major + 1}.0.0`;
    case "minor": return `${major}.${minor + 1}.0`;
    case "patch": return `${major}.${minor}.${patch + 1}`;
  }
}

// --- Changelog generation ---

export function generateChangelog(newVersion: string, commits: Commit[]): string {
  const date = new Date().toISOString().split("T")[0];
  const lines: string[] = [`## [${newVersion}] - ${date}`];

  const breaking = commits.filter((c) => c.breaking);
  const features = commits.filter((c) => !c.breaking && c.type === "feat");
  const fixes = commits.filter((c) => !c.breaking && c.type === "fix");
  const others = commits.filter((c) => !c.breaking && c.type !== "feat" && c.type !== "fix");

  function formatEntry(c: Commit): string {
    const prefix = c.scope ? `**${c.scope}**: ` : "";
    return `- ${prefix}${c.message} (${c.hash.slice(0, 7)})`;
  }

  if (breaking.length > 0) {
    lines.push("", "### Breaking Changes");
    breaking.forEach((c) => lines.push(formatEntry(c)));
  }
  if (features.length > 0) {
    lines.push("", "### Features");
    features.forEach((c) => lines.push(formatEntry(c)));
  }
  if (fixes.length > 0) {
    lines.push("", "### Bug Fixes");
    fixes.forEach((c) => lines.push(formatEntry(c)));
  }
  if (others.length > 0) {
    lines.push("", "### Other Changes");
    others.forEach((c) => lines.push(formatEntry(c)));
  }

  return lines.join("\n");
}

// --- File I/O ---

export function parseCommitsFile(commitsPath: string): Commit[] {
  if (!existsSync(commitsPath)) return [];
  const content = readFileSync(commitsPath, "utf-8").trim();
  if (!content) return [];
  return content.split("\n").filter((l) => l.trim()).map(parseCommit);
}

function readVersion(dir: string): { version: string; source: "package.json" | "version.txt" } {
  const pkgPath = join(dir, "package.json");
  const versionPath = join(dir, "version.txt");

  if (existsSync(pkgPath)) {
    const pkg = JSON.parse(readFileSync(pkgPath, "utf-8")) as { version: string };
    return { version: pkg.version, source: "package.json" };
  }
  if (existsSync(versionPath)) {
    return { version: readFileSync(versionPath, "utf-8").trim(), source: "version.txt" };
  }
  throw new Error("No version file found (package.json or version.txt)");
}

function writeVersion(
  dir: string,
  source: "package.json" | "version.txt",
  newVersion: string
): void {
  if (source === "package.json") {
    const pkgPath = join(dir, "package.json");
    const pkg = JSON.parse(readFileSync(pkgPath, "utf-8")) as Record<string, unknown>;
    pkg["version"] = newVersion;
    writeFileSync(pkgPath, JSON.stringify(pkg, null, 2) + "\n");
  } else {
    writeFileSync(join(dir, "version.txt"), newVersion + "\n");
  }
}

// --- Main bump function ---

export function bump(dir: string, commitsPath: string): BumpResult {
  const { version: oldVersion, source } = readVersion(dir);
  const commits = parseCommitsFile(commitsPath);
  const bumpType = determineBumpType(commits);
  const newVersion = bumpVersion(oldVersion, bumpType);
  const changelog = generateChangelog(newVersion, commits);

  if (bumpType !== "none") {
    writeVersion(dir, source, newVersion);

    const changelogPath = join(dir, "CHANGELOG.md");
    const existing = existsSync(changelogPath) ? readFileSync(changelogPath, "utf-8") : "";
    writeFileSync(changelogPath, changelog + "\n\n" + existing);
  }

  return { oldVersion, newVersion, bumpType, commits, changelog };
}

// --- CLI entry point ---

if (import.meta.main) {
  const dir = process.argv[2] ?? ".";
  const commitsPath = process.argv[3] ?? join(dir, "commits.txt");

  try {
    const result = bump(dir, commitsPath);
    console.log(`OLD_VERSION=${result.oldVersion}`);
    console.log(`NEW_VERSION=${result.newVersion}`);
    console.log(`BUMP_TYPE=${result.bumpType}`);
    console.log("");
    console.log(result.changelog);
  } catch (err) {
    console.error(`Error: ${(err as Error).message}`);
    process.exit(1);
  }
}

File: results/2026-05-06_173435/tasks/11-semantic-version-bumper/default-opus47-1m-medium/generated-code/bumper.py
"""
Semantic version bumper driven by Conventional Commits.

The flow is intentionally split into small pure functions so each one was easy
to test in isolation during the TDD cycle. The CLI / GitHub Action calls
``run`` which wires them together.
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from dataclasses import dataclass
from datetime import date as _date
from pathlib import Path
from typing import Iterable, List, Optional


class BumperError(Exception):
    """Raised for any user-facing failure: missing file, bad version, etc."""


@dataclass
class Commit:
    sha: str
    type: str
    scope: Optional[str]
    subject: str
    breaking: bool


# Conventional commit header: "type(scope)!: subject" with scope and ! optional.
_HEADER = re.compile(
    r"^(?P<sha>[0-9a-f]{4,40})\s+"
    r"(?P<type>[a-zA-Z]+)"
    r"(?:\((?P<scope>[^)]+)\))?"
    r"(?P<bang>!)?"
    r":\s*(?P<subject>.+?)\s*$"
)

_SEMVER = re.compile(r"^(\d+)\.(\d+)\.(\d+)$")


def read_version(path: Path) -> str:
    if not path.exists():
        raise BumperError(f"version file not found: {path}")
    try:
        data = json.loads(path.read_text())
    except json.JSONDecodeError as e:
        raise BumperError(f"{path} is not valid JSON: {e}") from e
    if "version" not in data:
        raise BumperError(f"{path} has no 'version' field")
    return data["version"]


def write_version(path: Path, new_version: str) -> None:
    data = json.loads(path.read_text())
    data["version"] = new_version
    path.write_text(json.dumps(data, indent=2) + "\n")


def parse_commits(log: str) -> List[Commit]:
    """
    Parse a git-log-ish blob.

    Expected layout per commit:
        <sha> <header line>
        <optional body lines>
        (blank line separates commits — also tolerated: each commit on its own line)

    Lines whose header doesn't match Conventional Commits are silently ignored
    so unrelated/legacy commits don't accidentally trigger a release.
    """
    # We walk lines top-to-bottom. A line matching the conventional header
    # starts a new commit; everything until the next header (or EOF) is body.
    commits: List[Commit] = []
    current: Optional[Commit] = None
    for raw in log.splitlines():
        line = raw.rstrip()
        m = _HEADER.match(line)
        if m:
            current = Commit(
                sha=m.group("sha"),
                type=m.group("type").lower(),
                scope=m.group("scope"),
                subject=m.group("subject"),
                breaking=bool(m.group("bang")),
            )
            commits.append(current)
            continue
        if current is not None and (
            line.lstrip().startswith("BREAKING CHANGE")
            or line.lstrip().startswith("BREAKING-CHANGE")
        ):
            current.breaking = True
    return commits


def determine_bump(commits: Iterable[Commit]) -> Optional[str]:
    has_feat = False
    has_fix = False
    for c in commits:
        if c.breaking:
            return "major"
        if c.type == "feat":
            has_feat = True
        elif c.type == "fix":
            has_fix = True
    if has_feat:
        return "minor"
    if has_fix:
        return "patch"
    return None


def bump_version(version: str, kind: str) -> str:
    m = _SEMVER.match(version)
    if not m:
        raise BumperError(f"invalid semantic version: {version!r}")
    major, minor, patch = (int(x) for x in m.groups())
    if kind == "major":
        return f"{major + 1}.0.0"
    if kind == "minor":
        return f"{major}.{minor + 1}.0"
    if kind == "patch":
        return f"{major}.{minor}.{patch + 1}"
    raise BumperError(f"unknown bump kind: {kind!r}")


def render_changelog(version: str, commits: List[Commit], date: Optional[str] = None) -> str:
    date = date or _date.today().isoformat()
    breaking = [c for c in commits if c.breaking]
    feats = [c for c in commits if c.type == "feat" and not c.breaking]
    fixes = [c for c in commits if c.type == "fix" and not c.breaking]
    lines = [f"## {version} - {date}", ""]
    if breaking:
        lines.append("### Breaking Changes")
        lines += [f"- {c.subject} ({c.sha})" for c in breaking]
        lines.append("")
    if feats:
        lines.append("### Features")
        lines += [f"- {c.subject} ({c.sha})" for c in feats]
        lines.append("")
    if fixes:
        lines.append("### Bug Fixes")
        lines += [f"- {c.subject} ({c.sha})" for c in fixes]
        lines.append("")
    return "\n".join(lines).rstrip() + "\n"


def _prepend(path: Path, block: str) -> None:
    existing = path.read_text() if path.exists() else ""
    path.write_text(block + ("\n" + existing if existing else ""))


def run(
    package_path: Path,
    commits_path: Path,
    changelog_path: Path,
    date: Optional[str] = None,
) -> dict:
    if not commits_path.exists():
        raise BumperError(f"commits file not found: {commits_path}")
    old = read_version(package_path)
    commits = parse_commits(commits_path.read_text())
    bump = determine_bump(commits)
    if bump is None:
        return {"old_version": old, "new_version": old, "bump": None, "commits": len(commits)}
    new = bump_version(old, bump)
    write_version(package_path, new)
    _prepend(changelog_path, render_changelog(new, commits, date=date))
    return {"old_version": old, "new_version": new, "bump": bump, "commits": len(commits)}


def main(argv: Optional[List[str]] = None) -> int:
    p = argparse.ArgumentParser(description="Semantic version bumper")
    p.add_argument("--package", default="package.json", help="path to package.json")
    p.add_argument("--commits", default="commits.txt", help="path to commit log fixture")
    p.add_argument("--changelog", default="CHANGELOG.md", help="path to changelog file")
    p.add_argument("--date", default=None, help="override date (YYYY-MM-DD) for reproducible output")
    args = p.parse_args(argv)
    try:
        result = run(
            package_path=Path(args.package),
            commits_path=Path(args.commits),
            changelog_path=Path(args.changelog),
            date=args.date,
        )
    except BumperError as e:
        print(f"error: {e}", file=sys.stderr)
        return 1
    # Machine-readable single-line summary, plus a friendly version line.
    print(f"old_version={result['old_version']}")
    print(f"new_version={result['new_version']}")
    print(f"bump={result['bump']}")
    print(f"commits={result['commits']}")
    print(f"VERSION={result['new_version']}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
